src/clj/rems/ega.clj

```diff
@@ -0,0 +1,259 @@
+  [entitlement]
+  {:format "JWT"
+   :visas (-> [{:resid (:resid entitlement)
+                :start (:start entitlement)
+                :end (:end entitlement)
+                :userid (:userid entitlement)
+                :dac-id (get-dac-for (:resid entitlement) (:approvedby entitlement))}]
+              ga4gh/entitlements->passport
+              :ga4gh_passport_v1)})
+
+(deftest test-entitlement->update
+  (with-redefs [env (rems.config/load-keys {:public-url "https://rems.example/"
+                                            :ga4gh-visa-private-key "test-data/example-private-key.jwk"
+                                            :ga4gh-visa-public-key "test-data/example-public-key.jwk"})
+                time-core/now (fn [] (time-core/date-time 2021 04 17))] ; iat
+    (is (= {:format "JWT"
+            :visas [[{:alg "RS256"
+                      :kid "2011-04-29"
+                      :jku "https://rems.example/api/jwk"
+                      :typ "JWT"}
+                     {:sub "elixir-user"
+                      :iss "https://rems.example/"
+                      :exp 1647475200
+                      :ga4gh_visa_v1
+                      {:value "EGAD00001006673"
+                       :type "ControlledAccessGrants"
+                       :source "EGAC00001000908"
+                       :asserted 1615939200
+                       :by "dac"}
+                      :iat 1618617600}]]}
+           (update-in (entitlement->update {:resid "EGAD00001006673"
+                                            :userid "elixir-user"
+                                            :approvedby "dac-user"
+                                            :start (time-core/date-time 2021 03 17)
+                                            :end (time-core/date-time 2022 03 17)})
+                      [:visas 0]
+                      ga4gh/visa->claims)))))
+
+(defn entitlement-push [action entitlement config]
+  (let [api-key (get-api-key-for (:approvedby entitlement))
+        common-fields {:api-key api-key
+                       :account-id (:userid entitlement)
+                       :config config}]
+
+    (when (str/blank? api-key)
+      (log/warnf "Missing EGA api-key for %s" (:approvedby entitlement)))
+
+    (log/infof "Pushing entitlements to %s %s: %s %s" (:id config) (:permission-server-url config) entitlement config)
+
+    (case action
+      :add
+      (post-create-or-update-permissions (merge common-fields
+                                                (entitlement->update entitlement)))
+
+      :remove
+      (delete-permissions (merge common-fields
+                                 {:dataset-ids [(:resid entitlement)]})))))
+
+(defn update-api-key
+  "Logs into EGA fetching an access token, generates an API-Key and saves it to the user's secrets.
+
+  `:userid`               - REMS user to generate the key for
+  `:username`             - EGA user to use
+  `:password`             - EGA user password
```

Comment:
hmm... weren't we supposed to lean on the users OIDC login instead of handling their password?

Reply:
This is for the other case, i.e. providing EGA user ids. The other one is actually not pushed yet I think.

Reply:
So for the case where the user ids don't match and you need to install the api-key manually? Like in the demo?

Reply:
This was mostly for the demo. The real use-case is that if you want to use the approver bot, you need a (CLI) command to fetch an API key. In that case I think you must use an EGA user and password and can't link an ELIXIR id to it (and use its token). 

There could be another process to use an ELIXIR user, get token, get API-key but let's do that only if needed. That would require an ELIXIR user participation and login, i.e., a handler that represents the approver bot.